torch/csrc/generic/Tensor.cpp

```diff
@@ -568,107 +625,270 @@ static PyObject* THPTensor_(checkAdvancedIndexing)(THPTensor *self, PyObject *ar
+
+    // Broadcast succeeded, place Broadcasted Tensors into output map by the index at
+    // which they occurred, transferring ownership to that map object
+    for (unsigned int i = 0; i < indexingDims.size(); ++i) {
+      THLongTensorPtr owned(maybeBroadcasted[i]);
+      broadcasted[indexingDims[i]] = std::move(owned);
+    }
+
+    // Next, before doing any further work, we want to verify that all the indices
+    // are in bounds at each advanced index dimension
+
+    ptrdiff_t nElement = THLongTensor_nElement(broadcasted.begin()->second.get());
+    THLongStoragePtr viewer(THLongStorage_newWithSize(1));
+    THLongStorage_set(viewer.get(), 0, nElement);
+    for (auto& dimBroadcast : broadcasted) {
+      Py_ssize_t dim = dimBroadcast.first;
+      long sizeAtDim = THTensor_(size)(LIBRARY_STATE indexed, dim);
+
+      // Need to make contiguous to view as 1D :/
+      THLongTensorPtr contig(THLongTensor_newContiguous(dimBroadcast.second.get()));
+
+      // View as 1D + get1D makes me sad :(
+      THLongTensorPtr flat(THLongTensor_newView(contig.get(), viewer));
+      for (ptrdiff_t i = 0; i < THLongTensor_nElement(flat.get()); ++i) {
+        long indexAtDim = THLongTensor_get1d(flat.get(), i);
+        if (indexAtDim >= sizeAtDim) {
+          PyErr_Format(PyExc_IndexError, "index %lld from broadcast indexer is out of range "
+              "for dimension %lld (of size %lld)",
+              (long long)indexAtDim, (long long)dim, (long long)sizeAtDim);
+          return false;
+        }
+      }
     }
   } catch (std::exception& e) {
-    // Broadcasted failed, cleanup and set error
-    for (int i = 0; i < seqCount; ++i) {
-      THLongTensor_free(maybeBroadcasted[i]);
+    // Broadcasted failed, cleanup and return error. I'm not sure if there is a better
+    // way to do this where we don't have to manually clean up the memory
+    for (const auto& tensor : maybeBroadcasted) {
+      THLongTensor_free(tensor);
     }
-    broadcastSuccess = false;
     PyErr_Format(PyExc_IndexError, "The advanced indexing objects could not be broadcast");
+    return false;
   }
 
-  // No matter what, need to cleanup the candidates
-  for (Py_ssize_t i = 0; i < seqCount; ++i) {
-    THLongTensor_free(candidates[i]);
+  // TODO: what do we do with candidates? Are the GC'ed, or do we need to do something with them?
+  return true;
+}
+
+static inline long THPTensor_(_indexToOffset)(
+    THTensorPtr& indexed,
+    std::unordered_map<Py_ssize_t, THLongTensorPtr>& broadcasted,
+    ptrdiff_t index)
+{
+  // We need to translate an "index" into a linear offset within the Tensor indexed.
+  // We will perform the normal mod/divide loop, except in the case of a broadcasted
+  // dimension, we need to take special care to utilize the broadcasted size and
+  // subset of indices
+  //
+  // For example, suppose we have a three-dimensional Tensor x of shape (5, 10, 15),
+  // and our indexing operation is x[:, (2, 4, 5), :].
```

Comment:
Yes, in this case, there is no "broadcasting" per-say, but we use broadcasted as a way to name a possibly-broadcasted advanced indexing Tensor. I can change the name to make it more clear.